src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -192,100 +242,157 @@ static void ResolveCallback(
     const unsigned char * txtRecord,
     void * context)
 {
-    auto * connectivityMonitor = (__bridge MTRDeviceConnectivityMonitor *) context;
-    [connectivityMonitor handleResolvedHostname:hostName port:port error:errorCode];
-}
+    // Context contains monitor ID, to look up the monitor object
+    NSUInteger monitorID = reinterpret_cast<NSUInteger>(context);
 
-- (void)startMonitoringWithHandler:(MTRDeviceConnectivityMonitorHandler)handler queue:(dispatch_queue_t)queue
-{
     std::lock_guard lock(sConnectivityMonitorLock);
+    MTRDeviceConnectivityMonitor *monitor = [sMonitorMap objectForKey:@(monitorID)];
 
-    _monitorHandler = handler;
-    _handlerQueue = queue;
-
-    // If there's already a resolver running, just return
-    if (_resolvers.size() != 0) {
-        MTR_LOG("%@ connectivity monitor already running", self);
+    // If monitor is nil, object was deallocated - callback safely ignored
+    if (!monitor) {
         return;
     }
 
-    MTR_LOG("%@ start connectivity monitoring for %@ (%lu monitoring objects)", self, _instanceName, static_cast<unsigned long>(sConnectivityMonitorCount));
+    [monitor handleResolvedHostname:hostName port:port error:errorCode];
+}
 
-    auto sharedConnection = [MTRDeviceConnectivityMonitor _sharedResolverConnection];
-    if (!sharedConnection) {
-        MTR_LOG_ERROR("%@ failed to get shared resolver connection", self);
-        return;
-    }
+- (BOOL)startMonitoringWithHandler:(MTRDeviceConnectivityMonitorHandler)handler queue:(dispatch_queue_t)queue
+{
+    // Initialize sSharedResolverQueue on first use
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sSharedResolverQueue = dispatch_queue_create("MTRDeviceConnectivityMonitor", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
+    });
+
+    __block BOOL result = NO;
 
-    for (auto domain : kResolveDomains) {
-        DNSServiceRef resolver = sharedConnection;
-        DNSServiceErrorType dnsError = DNSServiceResolve(&resolver,
-            kDNSServiceFlagsShareConnection,
-            kDNSServiceInterfaceIndexAny,
-            _instanceName.UTF8String,
-            kOperationalType,
-            domain,
-            ResolveCallback,
-            (__bridge void *) self);
-        if (dnsError == kDNSServiceErr_NoError) {
-            _resolvers.emplace_back(std::move(resolver));
-        } else {
-            MTR_LOG_ERROR("%@ failed to create resolver for \"%s\" domain: %" PRId32, self, StringOrNullMarker(domain), dnsError);
+    // Run entire function on sSharedResolverQueue to maintain lock ordering (queue → lock)
+    // and avoid deadlocks with callbacks that also use this pattern.
+    dispatch_sync(sSharedResolverQueue, ^{
+        std::lock_guard lock(sConnectivityMonitorLock);
```

Comment:
I guess the only places that touch the lock off the queue are dealloc implementations? 

But I'm not clear why we need to run on sSharedResolverQueue, given we have the lock.  Can't we just take the lock and then do all our work in the caller's threading context?  If anything tries to run on the queue it will take the lock before doing anything interesting....